src/librustc/infer/region_inference/mod.rs

```diff
@@ -1543,7 +1543,7 @@ impl<'a, 'gcx, 'tcx> GenericKind<'tcx> {
         match *self {
             GenericKind::Param(ref p) => p.to_ty(tcx),
             GenericKind::Projection(ref p) => tcx.mk_projection(
-                p.trait_ref.clone(), p.item_name(tcx)),
+                p.trait_ref(tcx).clone(), p.item_name(tcx)),
```

Comment:
`DefId` and `Substs`. No callsites should really need to search for the associated item *if they don't already*, e.g. anything to do with lang item traits with an associated type should just get the associated type *ignoring its name*.